### ISSUE TITLE
Does this implementation deprecate the boxes whose "iscrowd" values equals true while training on COCO dataset?

### ISSUE THREAD

**Issue by jackshaw** [facebookresearch/maskrcnn-benchmark#380]
## ❓ Questions and Help
I found that there is no extra processing steps if some anchors or proposals match to ground truth boxes whose "iscrowd" or "difficult" equal true when training. Does this mean that the implementation views bboxes whose "iscrowd" or "difficult"  equal true as negative samples? If I want to ignore the loss when matching to those special ground truth boxes, how should I do?

**Comment by looput** [comment]
I implement it in my own dataset. I add an extra field `difficult` (`is_crowd`) in the boxlist. In train period, First finding the proposals that match to `difficult` target, and labeling these proposals with ignore_value `-1` . In my own dataset, the result improves by 1~2%.
```bash
diff --git a/maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py b/maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py
index 991e86b..5582187 100644
--- a/maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py
+++ b/maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py
@@ -69,6 +69,16 @@ class FastRCNNLossComputation(object):
             # Label background (below the low threshold)
             bg_inds = matched_idxs == Matcher.BELOW_LOW_THRESHOLD
             labels_per_image[bg_inds] = 0
+            
+            # label ignore proposals which is difficult
+            difficult=targets_per_image.get_field('difficult')
+            diff_ids=torch.nonzero(difficult) # for those difficult labeled with 1
+            if diff_ids.shape[0]>0:
+                matched_idxs_re=matched_idxs.repeat([diff_ids.shape[0],1])  # [n_d,n_p]
+                diff_ids_re=diff_ids.view(-1,1).repeat([1,matched_idxs.shape[0]])
+                diff_match_matrix=matched_idxs_re==diff_ids_re     # [n_d,n_p]
+                ignore_value,ignore_ids=diff_match_matrix.max(dim=0)  # [n_p]
+                labels_per_image[ignore_value]=-1
```
In some case, if we filter the difficult samples, the positive proposal may be empty, so I handle this situation by code below. 
```bash
diff --git a/maskrcnn_benchmark/modeling/roi_heads/mask_head/mask_head.py b/maskrcnn_benchmark/modeling/roi_heads/mask_head/mask_head.py
index c8c8b08..f80b5fc 100644
--- a/maskrcnn_benchmark/modeling/roi_heads/mask_head/mask_head.py
+++ b/maskrcnn_benchmark/modeling/roi_heads/mask_head/mask_head.py
@@ -64,6 +64,8 @@ class ROIMaskHead(torch.nn.Module):
             # during training, only focus on positive boxes
             all_proposals = proposals
             proposals, positive_inds = keep_only_positive_boxes(proposals)
+            if all([len(proposal)==0 for proposal in proposals])==True:
+                return None,None,None
         if self.training and self.cfg.MODEL.ROI_MASK_HEAD.SHARE_BOX_FEATURE_EXTRACTOR:
             x = features
             x = x[torch.cat(positive_inds, dim=0)]
@@ -75,7 +77,14 @@ class ROIMaskHead(torch.nn.Module):
             result = self.post_processor(mask_logits, proposals)
             return x, result, {}

-        loss_mask = self.loss_evaluator(proposals, mask_logits, targets)
+        proposals_not_empty,targets_not=[],[]
+        for proposal_target in zip(proposals,targets):
+            if len(proposal_target[0])>0:
+                proposals_not_empty.append(proposal_target[0])
+                targets_not.append(proposal_target[1])
+        
+        # loss_mask = self.loss_evaluator(proposals, mask_logits, targets)
+        loss_mask = self.loss_evaluator(proposals_not_empty, mask_logits, targets_not)

         return x, all_proposals, dict(loss_mask=loss_mask)

diff --git a/maskrcnn_benchmark/modeling/roi_heads/roi_heads.py b/maskrcnn_benchmark/modeling/roi_heads/roi_heads.py
index c01747d..7dfe7e0 100644
--- a/maskrcnn_benchmark/modeling/roi_heads/roi_heads.py
+++ b/maskrcnn_benchmark/modeling/roi_heads/roi_heads.py
@@ -39,7 +39,8 @@ class CombinedROIHeads(torch.nn.ModuleDict):
             if targets is not None:
                 if all([len(target)==0 for target in targets])==False:
                     x, detections, loss_mask = self.mask(mask_features, detections, targets)
-                    losses.update(loss_mask)
+                    if loss_mask is not None:
+                        losses.update(loss_mask)
```

**Comment by Tiamo666** [comment]
@looput 
Hello,  your solution is great, that's a good job. 
And I think we should also ignore difficult samples when computing RPN loss. Cause they should not be considered as foreground or background.

**Comment by looput** [comment]
@Tiamo666 I think so! On the other if we treat difficult samples as positive, would this increase recall？I haven't experimented yet.

**Comment by hongzhenwang** [comment]
@looput I change the code according to your suggestion. But, training procedure will hang on sometimes. Have to kill the code. Did you have this problem?
![企业微信截图_15584256056782](https://user-images.githubusercontent.com/9897079/58081111-0e5abc00-7be7-11e9-871a-db30a940db68.png)

**Comment by Tiamo666** [comment]
@looput In my own dataset, it does increase the recall

**Comment by nnop** [comment]
Should we trean the `iscrowd` boxes as negative or ignore samples?
